### ISSUE TITLE
Clarification: Should a CNR be sent in case "we have something -> we have nothing"

### ISSUE THREAD

**Issue by j-be** [erasmus-without-paper/ewp-specs-api-omobility-la-cnr#2]
Imagine the following:

* LA is created -> no notification, as LA shall not be exposed to EWP yet
* LA is edited, but not signed -> no notification, as LA shall not be exposed to EWP yet
* LA is signed by both Student and sending HEI -> CNR is sent to receiving HEI
* Somebody at sending HEI notices an error in LA, which requires it to be changed -> both signatrues expire, LA shall not be exposed over EWP until it is signed again

Should I now trigger a CNR to receiving HEI? From spec's `whenever learning agreements kept on their partner institutions' servers are changed` but without being able to pass on a reason (like `gone` or `retracted` or something along that line) the only thing I get is an invalid request of the HEI when trying to fetch the update, as there is no LA anymore for the `omobility_id` I'm posting.

EDIT: After reading through the documentation another couple of times, I think there is deeper misunderstanding on my side: As I am required by https://github.com/erasmus-without-paper/ewp-specs-api-omobility-las/blob/67a63dc6fef3eacab1af7eff19144817ad57cc43/endpoints/get-response.xsd#L583-L584 to have `student-signature`and `sending-hei-signature` I thought to only expose LAs if they have either `changesPropsal` with the 2 signatures, or `firstVersion` set. But that would require me to be able to "unpublish" LAs as described above. I don't see anything in spec preventing me from doing that, but is it allowed?

**Comment by kamil-olszewski-uw** [comment]
Yes, in the case you describe you should "unpublish" LA. And in this case receiving HEI will get an answer with invalid request info.

We realize that this is not an elegant solution, but it must suffice for a while. Similiar situation will arise when LAs are removed. The solution we will develop will also fit the new LA (without first version signed by all three parties) withdrawn for corrections.

**Comment by j-be** [comment]
Ok, so just to be sure (as that has influence on incoming implementation) in case a `changes-propsal` is withdrawn for correction, I MUST do one of the following:

* "unpublish" the LA as a whole
* keep the LA public, but with none of `changes-propsal`, `first-version`, nor `approved-changes`

In either case I MUST send a CNR to receiving HEI.

As a receiving HEI I MUST take into account, that when receiving a CNR I may not get an LA for the given `omobility-id` (which kind of breaks referential integrity). Additionally, I MUST take into account, that a LA may just be an empty shell without neither of `changes-propsal`, `first-version`, nor `approved-changes`.

Taking all of this one step further, I MAY even publish an empty LA shell (EDIT: i.e. only the required fields) as soon as one is created in our system (i.e. without any signature on it yet, so neither of `changes-propsal`, `first-version`, nor `approved-changes` is present), then I:
* MUST NOT send CNRs as long as the `changes-propsal` are not public
* MUST send a CNR if `changes-propsal` becomes public
* MUST send a CNR if `changes-propsal` changes while they are public (imho. not actually possible assuming the current LA spec)
* MUST send a CNR if the LA goes back to being an empty shell

Is that correct or is anything of the above prohibited by spec?

And about the meta-state: changing a LA in any way would void any signature currently present, right? So actually a `changes-propsal` can't change while public, because - even if a change implies signing - after a change there is at least one signature missing (i.e. supervisor makes the change -> student needs to re-sign, and vice-versa). So a CNR can only mean either of
* we had some `changes-propsal` for you, now we don't
* we didn't have `changes-propsal` for you, now we have some

Is that correct?

EDIT: While thinking through it once again: Am I required to send a CNR in case Receiving HEI signs? Technically, that changes the LA, practically it makes very little sense to do this.

**Comment by kamil-olszewski-uw** [comment]
There is no empty shell sent via API. LA must always have at least first-version or changes-proposal.

If agreement without first-version requires corrections on the sending HEI's side, it temporarily disappears from the EWP.

In the new LA, you must not send CNRs as long as the changes-proposal are not public. You must send a CNR if changes-proposal becomes public.

If you want to make changes to changes-proposal then you need to create new changes-proposal (signed by student and sending HEI) with new hash id. You can hide old changes-proposal and publish new changes proposal later. In both cases, if receiving HEI will send update to old changes-proposal (with old hash id), sending HEI will send answer 409 (conflict).

Not every change voids signatures. There are minor changes (e.g. contact details) which do not create a new LA version.

CNR is not only used to reveal or hide changes-proposal. For example, it also tells you that there has been minor changes or that update with approval of changes-proposal has been processed and a new LA version has been created.

**Comment by j-be** [comment]
> There is no empty shell sent via API. LA must always have at least first-version or changes-proposal.

Imho. that is not what the spec says. Both `first-version` and `changes-proposal` are `minOccurs="0" maxOccurs="1"`. And I don't see any other restriction in [get-response.xsd, v0.4.0](https://github.com/erasmus-without-paper/ewp-specs-api-omobility-las/blob/v0.4.0/examples/get-response-example.xml) preventing me from doing that. So technically I (or anybody else) am allowed to publish such empty shells with only the bare minimum required fields.

> For example, it also tells you that there has been minor changes or that update with approval of changes-proposal has been processed and a new LA version has been created.

Now that complicates things... That means I need to send a CNR if e.g. there was a typo in `sending-hei.ounit-name`. But the receiver has no possibility to find out what exactly changed. Additionally, EWP basic architecture already states, that you must never trust your caches, if you have some.

In other words: if I don't expect any reaction from receiving HEI, why should I notify them?

> that update with approval of changes-proposal has been processed

Are you impling, that "Outgoing Mobility Learning Agreements Update endpoint" is not necessarily synchronous? In other words: do I need to take care of situations, where we as receiving university signed (i.e. with the right `id` and everything, and I got a `200` from the endpoint), but the signature was either lost, or is just heavily delayed for some reason?

**Comment by kamil-olszewski-uw** [comment]
> > There is no empty shell sent via API. LA must always have at least first-version or changes-proposal.
> 
> Imho. that is not what the spec says. Both `first-version` and `changes-proposal` are `minOccurs="0" maxOccurs="1"`. And I don't see any other restriction in [get-response.xsd, v0.4.0](https://github.com/erasmus-without-paper/ewp-specs-api-omobility-las/blob/v0.4.0/examples/get-response-example.xml) preventing me from doing that. So technically I (or anybody else) am allowed to publish such empty shells with only the bare minimum required fields.

You are right that it is not clear in the specification. Thanks for pointing that out. We will consider releasing a hotfix to the API.

> > For example, it also tells you that there has been minor changes or that update with approval of changes-proposal has been processed and a new LA version has been created.
> 
> Now that complicates things... That means I need to send a CNR if e.g. there was a typo in `sending-hei.ounit-name`. But the receiver has no possibility to find out what exactly changed. Additionally, EWP basic architecture already states, that you must never trust your caches, if you have some.

Yes, CNR never tells you what has been changed. Receiving HEI needs to check if there is a new proposal, a new LA version or a new minor changes.

There will be no quick fixing of the typos. If there is a typo that needs to be corrected, we need to create a new changes-proposal with deletion and addition of component.

> In other words: if I don't expect any reaction from receiving HEI, why should I notify them?

The general idea is that CNR (in any API) should be sent with every change, and there are no exceptions to this.

> > that update with approval of changes-proposal has been processed
> 
> Are you impling, that "Outgoing Mobility Learning Agreements Update endpoint" is not necessarily synchronous? In other words: do I need to take care of situations, where we as receiving university signed (i.e. with the right `id` and everything, and I got a `200` from the endpoint), but the signature was either lost, or is just heavily delayed for some reason?

We cannot assume that after receiving the code 200, the new version of LA automatically comes into effect. The only evidence is the relevant information in get response, preceded by CNR.

**Comment by j-be** [comment]
On a different topic, after reading through CNR specs once again: a lot of places seem to dance around "if you offer a CNR API endpoint for `EntityX`, you MUST also send CNRs for `EntityX`", but nowhere does it seem to explicitly state that. I find this especially important for this CNR as it is required by the LA API spec. Is this true, or am I allowed to implement a stub LA CNR endpoint (assuming I don't need the information) and not send out CNRs for LA changes?

> We cannot assume that after receiving the code 200, the new version of LA automatically comes into effect. The only evidence is the relevant information in get response, preceded by CNR.

Can we assume the new version as signed will *eventually* come into effect if we get a `200` on `update`? Or is there any kind of race-condition like "we get `200` on an asyncronous update, but a contemporary major change at sending HEI takes precedence, nullifying our signature?" the spec does not cover, hence the client has to deal with it?

Being a free-text message, parsing the response on an `update` is not really an option.

**Comment by janinamincer-daszkiewicz** [comment]
Ad. On a different topic
All CNRs should be treated as 'a must'. Eventually our new Registration portal (we are designing now) will be verifying if all APIs from the same 'family' are available and if not will omit them. For example for IIA all 5 might be expected.

**Comment by kamil-olszewski-uw** [comment]
> > We cannot assume that after receiving the code 200, the new version of LA automatically comes into effect. The only evidence is the relevant information in get response, preceded by CNR.
> 
> Can we assume the new version as signed will _eventually_ come into effect if we get a `200` on `update`? Or is there any kind of race-condition like "we get `200` on an asyncronous update, but a contemporary major change at sending HEI takes precedence, nullifying our signature?" the spec does not cover, hence the client has to deal with it?

Code 200 cannot be a proof of signing a new version. We can't assume anything about the implementation on the partner university side - it may even be that partner approvals are confirmed manually by the user, and code 200 only confirms that the answer was given to the still valid changes-proposal and the approval was put into the event queue.

**Comment by j-be** [comment]
> All CNRs should be treated as 'a must'.

@janinamincer-daszkiewicz That is in complete contradiction with pretty much every single point detailed in https://github.com/erasmus-without-paper/ewp-specs-architecture#cnr

Additionally, my question was: *IF* I offer a CNR API endpoint, am I required to send out CNR to other institutuions as well, or can I just *receive* CNRs (because I have to), but never ever *send* them (because I don't want to)? The spec does NOT seem explicitly state that I have to *send* any CNR, ever, no matter what. So, as far as I understand the `LearningAgreemet` specs as they are, I am required to *receive* CNRs (though I am allowed to just return a hardcoded `200` with the right XML and not do anything with the information), but I am not required to *send* CNRs if entities at my side change. Is that correct? If not, where does the spec require me to do so?

> Code 200 cannot be a proof of signing a new version

I see. That is the exact opposite of what I expected. Please consider making that clear in the spec. Additionally, I suggest to specify different status codes for the 2 cases:

* `202` means "acknoleged, needs to be processed" (i.e. asyncronous API, subsequent GET may not return a fully signed LA)
* `200` means "signature was successfully added to LA" (i.e. the syncronous version, subsequent GET returns a fully signed LA)

This could come in handy when developing the client for it, and there is not much effort involved on server side implementation.

**Comment by janinamincer-daszkiewicz** [comment]
Have a look at https://github.com/erasmus-without-paper/ewp-specs-architecture/issues/30 and you will see that this was not our decision but the majority. The specification will be slowly adapted to this decision.

You may behave in the network as you wish and implement as you wish with the end result that your partners find out and will not be willing to cooperate with you.

**Comment by j-be** [comment]
@janinamincer-daszkiewicz thanks for the hint with erasmus-without-paper/ewp-specs-architecture#30 - I was under the assumption `master` branch would contain a note if architecture fundamentals are about to be turned into their exact opposite. May I suggest adding such a note?

> find out and will not be willing to cooperate with you

I think I disagree with you. Since we all will be required to use EWP, as long as everybody behaves within spec - and that includes that everybody is free to opt for any interpretation of any ambiguity in the spec as sHe pleases - I don't think there is much ground for *not be willing to cooperate*. Of course big players can bully around the little ones, and force their own interpretation onto them. But I (representing one of the little players) really think that should be avoided by leaving as little gray areas in the spec as possible. Even further, the more gray areas, the more different interpretations are allowed, the more corner cases need to be handled, the more complex every single implementation needs to be.

I am merely pointing out here, that I think there is a gray area in the spec, which I think should be clarified.
